### ISSUE TITLE
Add cy.hover()

### ISSUE THREAD

**Issue by jennifer-mann** [cypress-io/cypress#10]
Need to be able to hover in/out of DOM elements.

##### Example for hover
```
cy.get("@firstRow").hover()
```

I'm not sure what a stop hover command would look like. 

Currently I would achieve this with `mouseOver` and `mouseOut` events. The hover command could follow a similar convention.

**Other considerations:**
- In what direction does the mouse come from for hover
- How long should the hover occur.
- How do you "pause" a hover to allow for possible css animations.
- Which actions cause you no longer to hover.

**Comment by jennifer-mann** [comment]
When I issue a `.click()` command, this should issue a `.hover()` over the element I've specified to click onto (in order to fully simulate mouse behavior).

**Comment by seltar** [comment]
I don't think it's possible to implement a hover function with support for the css pseudo selectors considering it's a trusted event. http://www.w3.org/TR/DOM-Level-3-Events/#trusted-events
The only way to simulate a hover to play css animations is by adding / removing a class. 

A solution could be to modify and reinject the css to ensure that all :hover pseudo selectors are made available as a class.

**Comment by brian-mann** [comment]
Correct, it is not possible to trigger the CSS pseudo selectors via JavaScript.

When browsers run in cross browser mode we simply use the OS to actually hover over an element by its coordinates, so this isn't an issue in that mode.

However when testing in your local environment, Cypress has to simulate all events, and thats where the problem comes in.

I experimented with a few different solutions nearly a year ago, and found one to work really well. It works like this - on `hover` find the element at the center of the coordinates, walk up the parent tree for every element we are currently hovering based on the CSS box model (not the DOM object model).

For each of those elements iterate through the css and figure out if a hover pseudo state applies. If it does, then inline that style directly on each element.  This will achieve all CSS results in JavaScript.

However that's only half the battle.

The reason this command hasn't been finished is 2 reasons:

1. Figuring out the parents by the CSS box model and not the DOM took awhile to get right - but now it is correct and it's already been implemented by `cy.click` - so this is no longer a barrier (but it used to be)
2. There are other things that go with a "hover" besides just CSS styles. For instance: `mouseover`, `mouseout` need to be simulated as well (which is easy at this point).  But the biggest issue here is semantics. For instance, at what point does the user **stop** hovering over an element? Deciding *when* to fire the `mouseout` is the real challenge.

I originally imagined `cy.hover` to actually be a callback function like this:

```javascript
cy.get("button").hover(function(){
  cy.get("span").click()
  ...
  ...
})
```

However that puts a lot of work on the user, and its possible you can create impossibilities. But the idea was that by using a callback function the user could specify the semantics of "continue hovering until X".

That is probably overkill though.  Instead I will just dumb down the method and if edge cases occur handle them when they arise.

Hover additionally needs to be automatically invoked under the hood to properly simulate user action.  For instance any `click` should automatically first hover over the element, and I guess `mouseout` when the action is done?

Anyway, that's why `hover` hasn't been done yet. It was dependent on a lot of other commands to be figured out, although it is ready now to be solved.

**Comment by jennifer-mann** [comment]
I hope to be able to write assertions about pseudo-classes since this isn't readily accessible without the work outlined above by @brian-mann. 

**Example assertion**
```
cy.get(“.content”).hover().should("have.attr", "style”, “cursor:alias”)
```

`.hover()` should return element chained to hover with the element having applied the styles specified within `:hover` pseudo-class.

**Comment by mlunoe** [comment]
This helped me out to at least show my element to be able to click it:
https://docs.cypress.io/v1.0/docs/invoke#section-properties-that-are-functions-are-invoked

```js
cy.get('.content').invoke('show').click();
```

**Comment by brian-mann** [comment]
Good point. We're going to add a pivotal story to write the documentation for `cy.hover` which redirects you and provides examples how to easily solve this.

**Comment by jennifer-mann** [comment]
Using cy.hover() will now provide a detailed error message with a link for working around hover constraints in [`0.15.0`](https://github.com/cypress-io/cypress/wiki/Changelog#0150-03282016).

**Comment by nazar** [comment]
The following workaround works for me:

```
Cypress.addChildCommand('triggerHover', function(elements) {

  elements.each((index, element) => {
    fireEvent(element, 'mouseover');
  });

  function fireEvent(element, event) {
    if (element.fireEvent) {
      element.fireEvent('on' + event);
    } else {
      var evObj = document.createEvent('Events');

      evObj.initEvent(event, true, false);

      element.dispatchEvent(evObj);
    }
  }
  
});
```

Which now allows me to:

```
          cy
            .get('.resultsHeading .icon-info-circle:first')
            .triggerHover()
            .get('.header-tooltips')
            .should('exist') 
  ```

HTH someone.

**Comment by brian-mann** [comment]
Yup, this simulates the event which is likely good enough in situations where you only have a simple javascript bound event handler.

To make this solution more complete we would need to:

1. Fill in the missing event coordinates such as `clientX/Y` properties which tell where the event is originating on the page. We calculate these for other Cypress commands such as `cy.click` which by default uses the exact center of the element, but this is configurable.

2. Replicate handling the CSS object model - meaning the event should actually originate at the topmost element (which is how we handle `cy.click` as well).

3. Still the hardest thing about hover is that you cannot invoke CSS pseudo properties from Javascript. There is no way to trigger the `:hover` styles on an element, which is likely what you're trying to do with `hover`. The only way to get around this is to go outside of the Javascript sandbox and use browser drivers aka the debugger protocol to fire native events - or parse all of the applied `:hover` rules and invoke them manually by inlining their styles on the element. Inlining styles will work but it has some edge cases. Creating a native event also has drawbacks (you can't have dev tools open, etc).

My thoughts are we still attempt to workaround hover purely in Javascript by inlining styles, but provide a `{native: true}` option which fires a native event. Once we do this, we can enable this option for all of the other commands too because.

I have on good authority been told that dev tools will eventually support multiplexed connections which means we'd no longer being booted off of the protocol whenever dev tools is opened by the user. This would make generating native events the preferred way in Cypress as opposed to event simulation.

**Comment by paulfalgout** [comment]
Re the example above... for javascript hovers this works:
```js
cy
  .get('.resultsHeading .icon-info-circle:first')
  //.triggerHover()
  .then(function($heading) { $heading.trigger('hover'); })
  .get('.header-tooltips')
  .should('exist') 
```

**Comment by brian-mann** [comment]
As per the above example - this will only work if you're using `jquery` to bind the `hover` event, since this is calling the `trigger` method on the jquery element.

Also you can write what you wrote a bit more succinctly like this:

```javascript
cy
  .get('.resultsHeading .icon-info-circle:first')
  .invoke('trigger', 'hover')
  .get('.header-tooltips')
```

Lastly you can ditch the `should('exist')` because that is the default assertion on all DOM commands.

https://docs.cypress.io/docs/finding-elements#section-existence

**Comment by brian-mann** [comment]
We've added a recipe which explains how to simulate hover / get around this.

https://github.com/cypress-io/cypress-example-recipes/blob/master/cypress/integration/hover_hidden_elements.js

**Comment by guoyang007** [comment]
I simulate the hover by dispatching the event 'mousemove',and it worked for me.
the question I faced with is : the chart is painted by SVG based on the React framework, and the tooltip can be seen only when the mouse hover it .

my solution is 
```
// create a event 'mousemove'
let e = new Event('mousemove', { bubbles: true, cancelable: true })

// give corresponding value to its target position
e.clientX=300;e.clientY=400

// dispatch the event
selector.dispatchEvent(e)
```

**Comment by brian-mann** [comment]
With `cy.trigger()` you can do things like...

```js
cy.get('button').trigger('mouseover') // or mouseenter based on what your app is bound to
```

**Comment by NateDukatz** [comment]
@brian-mann I'm curious if you've seen any issues using .trigger('mouseover') with php/jqery elements? I am unable to make it work. It will go to the correct element but the element doesn't recognize it as a mouse hover.

**Comment by brian-mann** [comment]
Events all work the same way in Javascript. As long as you trigger the event that your element/plugin/whatever is bound to - it will fire.

It's possible its not bound to `mouseover`, or that you'll need to provide additional properties to make it work. It all completely depends on how your code is written.

**Comment by NateDukatz** [comment]
Apparently this is what we are using:
https://jolt.slack.com/archives/D80V7N5G8/p1512155731000227

**Comment by NateDukatz** [comment]
I guess hoverIntent is what we are using:
https://github.com/briancherne/jquery-hoverIntent

**Comment by paulfalgout** [comment]
Regarding this issue we finally ran into a scenario in which we have a `:hover` style that sets the `display` and it can't be reproduced currently in JS.  Rather than adding JS to handle the situation we just added a duplicate style for `.cy-hover` and we `addClass` that class in the cypress test itself via a custom command.  This isn't something I'd want to get in the habit of, but it seems like a fairly unique scenario and this prevents the need for ad-hoc js hover solutions when a `:hover` will do.

**Comment by brian-mann** [comment]
Understood. Any reason you couldn't just force the display of the element itself as opposed to modifying its CSS styles?

Something like: `cy.get('button').invoke('show')`

**Comment by paulfalgout** [comment]
Hmm yeah in the cases where the `:hover` is only affecting the `display` or rather the only thing needed for the test that would cover it, but you'd lose any transitions or whatever else might be on the style

**Comment by cheahkhing** [comment]
I tried different suggestions provided but still find it hard to simulate the correct stylings that should happen on all the elements with css :hover, which is a pain.
If there are a lot of css :hover styles implemented, you'll see the application renders very differently from the actual visual. :(

**Comment by cheahkhing** [comment]
in fact, seeing how puppeteer can support hovering with their [hover](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#framehoverselector) i still think this is possible as well for cypress. :/

**Comment by maheswaranunni** [comment]
I am trying to automate one hover event in an angular application. Below is the code..

 <div class="childAvaName boxSize auto_test_dp parentPart ui-listHeader ui-collapse-button ng-scope handClick" ng-class="{'handClick':(parent.children.length > 2)}" ng-include="family_dropdown_obj.parent_dropdown_path" ng-init="dropdown_mode = 'header'" ng-mouseleave="bigListShow(false)" ng-mouseover="bigListShow(true)"><div class="childAva img-circle bgFit ng-scope" id="auto_test_dp" ;);">


I tried cy.get('.auto_test_dp').trigger('mouseover')
But after mouseover, the dropdown modal is not showed. 
I had tried many options and none of them are working. can u suggest any other options instead of doing force:true. ?

**Comment by brian-mann** [comment]
@maheswaranunni you just need to look at your application's code in response to the `mouseover` event. It likely is doing some math on `clientX` or `clientY` properties. If that's the case you'll need to provide those.

**Comment by stubar** [comment]
The :hover problem is interesting.  How about iterating through 
document.styleSheets finding ```:hover``` in  ```.cssText```.  If found make and attach a copy of it but switching :hover for some arbitrary fake hover class e.g. .cypress-fake-hover.  Then add that class in js when testing.

**Comment by novarichuffman** [comment]
Would it not be easy to just make a custom command that uses prevSubject for an element than too just say subject.trigger('mouseover') so the command would be cy.hover() right after you get an element?

**Comment by stubar** [comment]
I'm not sure I follow, would that pseudo css :hover effects?

**Comment by dukedave** [comment]
If you are a [Bulma](https://bulma.io/) user then this works in lieu of the CSS [`:hover`](https://bulma.io/documentation/components/dropdown/#hoverable-or-toggable) implementation:

```
      cy.get("#my-dropdown").within(() => {
        cy.get(".is-hoverable").invoke("addClass", "is-active");
      });
```

**Comment by prashantabellad** [comment]
I am using tippy.js to display tool tips. .trigger('mousehover') does not work.
Please suggest  a way

**Comment by monysaku** [comment]
Hi guys! We have an AngularJS application and we are using cypress for E2E tests. However we got stuck in a problem with hovering over elements. In particular we have a side menu with some icons. Hovering on an icon will trigger the panel with side menu links. Writing the test with cypress we find out that hover is not simple as it looks. In particular for this menu we are using angular helpers like ng-mouseenter to trigger hover and add the class .open that will trigger the side panel with the side menu links. Debugging we found out that the system find the element but nothing can trigger the mouseover on it. Suggestions here https://docs.cypress.io/api/commands/hover.html#Workarounds weren’t effective. We already used the dispatchEvent method without any success. Any suggestions? Thanks in advance for your support.

**Comment by DavidGWheeler** [comment]
> The following workaround works for me:
> 
> ```js
> Cypress.addChildCommand('triggerHover', function(elements) {
> 
>   elements.each((index, element) => {
>     fireEvent(element, 'mouseover');
>   });
> 
>   function fireEvent(element, event) {
>     if (element.fireEvent) {
>       element.fireEvent('on' + event);
>     } else {
>       var evObj = document.createEvent('Events');
> 
>       evObj.initEvent(event, true, false);
> 
>       element.dispatchEvent(evObj);
>     }
>   }
> 
> });
> ```
> 
> Which now allows me to:
> 
> ```
>           cy
>             .get('.resultsHeading .icon-info-circle:first')
>             .triggerHover()
>             .get('.header-tooltips')
>             .should('exist') 
> ```
> 
> HTH someone.

Is any of this actually functional in the latest version of Cypress?  I'm using TS by the way, but things like "addChildCommand" and the fact that your function expects an element but your example shows you passing none seem quite wrong.  Am I missing something?

**Comment by jennifer-shehane** [comment]
@DavidGWheeler 	`Cypress.addChildCommand` was replaced with `Cypress.Commands.add`. 

We prioritize issues we work on based on a variety of factors - including the number of users requesting/helped by a feature. While this is still on our Roadmap, many other features are a higher priority today.

**Comment by lukenofurther** [comment]
The `.trigger('mouseover')` solutions don't cause CSS changes, so we are left without a workaround even.

**Comment by inorganik** [comment]
Agree with @lukenofurther, the trigger workaround does not work. I can call `click({ force: true })`, and click a button that isn't visible but I can't assert that my UI is showing the button when an element is hovered.

**Comment by nym** [comment]
Given that this is a permission issue with trusted events, is the Firefox or Chome teams willing to help get this working through command line args like --untrustedEventsTrusted --Iknowitsaterribleidea.

@paulirish batsignal.jpg 📡

**Comment by nym** [comment]
A potential cypress fix might be to make fake :hover's in the transform step, e.g. change `div:hover { display: block }` to `div:hover, div.cy-hover { display: block }` 

Would of course like a better solution. This is a big pain point in testing visual changes.

**Comment by stubar** [comment]
There is no way of triggering psuedo selectors with javascript. I think my
idea of iterating CSS is the only way.

**Comment by ari1988** [comment]
> We've added a recipe which explains how to simulate hover / get around this.
> 
> https://github.com/cypress-io/cypress-example-recipes/blob/master/cypress/integration/hover_hidden_elements.js

the link throws 404 error

**Comment by warpdesign** [comment]
> > The following workaround works for me:
> > ```js
> > Cypress.addChildCommand('triggerHover', function(elements) {
> > 
> >   elements.each((index, element) => {
> >     fireEvent(element, 'mouseover');
> >   });
> > 
> >   function fireEvent(element, event) {
> >     if (element.fireEvent) {
> >       element.fireEvent('on' + event);
> >     } else {
> >       var evObj = document.createEvent('Events');
> > 
> >       evObj.initEvent(event, true, false);
> > 
> >       element.dispatchEvent(evObj);
> >     }
> >   }
> > 
> > });
> > ```
> > 
> > 
> > Which now allows me to:
> > ```
> >           cy
> >             .get('.resultsHeading .icon-info-circle:first')
> >             .triggerHover()
> >             .get('.header-tooltips')
> >             .should('exist') 
> > ```
> > 
> > 
> > HTH someone.
> 
> Is any of this actually functional in the latest version of Cypress? I'm using TS by the way, but things like "addChildCommand" and the fact that your function expects an element but your example shows you passing none seem quite wrong. Am I missing something or is this just outdated? If the latter is true can someone help me understand what changes are needed?
> 
> P.S. why is this issue still lingering after nearly 4 years of talking about it???

Did you find a way to make the workaround work with TypeScript? I am having the same problem.

**Comment by syn-zeta** [comment]
Haven't seen this mentioned anywhere, but the package [Mr0grog/nightmare-real-mouse](https://github.com/Mr0grog/nightmare-real-mouse) worked well for me when I was automating something that required hover action.

The implementation may be basic, but if it fulfills majority of the simple use cases, then it may be worth porting it over.

**Comment by avallete** [comment]
As mentioned above, puppetter does handle element hovering pretty well.

Digging onto the code to search how he's doing it, I found something who look like an elegant solution to avoid a lot of troubles who come from dealing with the DOM and CSS styles.

The hover function is defined like so:

```js
  async hover() {
    await this._scrollIntoViewIfNeeded();
    const {x, y} = await this._clickablePoint();
    await this._page.mouse.move(x, y);
  }
...
// _page.mouse.move
  async move(x, y, options = {}) {
    const {steps = 1} = options;
    const fromX = this._x, fromY = this._y;
    this._x = x;
    this._y = y;
    for (let i = 1; i <= steps; i++) {
      await this._client.send('Input.dispatchMouseEvent', {
        type: 'mouseMoved',
        button: this._button,
        x: fromX + (this._x - fromX) * (i / steps),
        y: fromY + (this._y - fromY) * (i / steps),
        modifiers: this._keyboard._modifiers
      });
    }
  }
```

So, basically, puppeter leverage the `Input.dispatchMouseEvent` from the devtools and simply move the mouse to the desired hover position. I've seen into cypress code that CDP's are already used into the `browsers` section.

I am curious to know if this kind of implementation would look like a decent solution to the maintainers, also, I think this is related to #311 since this approach do trigger native events ([example](https://puppeteersandbox.com/9VcGZU20))

**Comment by gabbersepp** [comment]
For those who are looking for a solution to trigger ":hover": Since cypress 3.8.0 you can run chrome in headless mode (there was an open issue that has been fixed and is available in the next release, see: https://github.com/cypress-io/cypress/issues/5949#issuecomment-567735351 )

So you can utilize the chrome remote debugger protocol. See this example: https://github.com/cypress-io/cypress-example-recipes/pull/384

**Comment by ericjorgeamaral** [comment]
For me this command does not work:
`cy.get('button').trigger('mouseover') // yields 'button'`
----------------------------------------------------------------------
as a solution I have done:
`cy.get('button').trigger('mouseenter')`
and then
`cy.get('button').trigger('mouseleave')`
------------------------------------------------------------------------

**Comment by Tay08** [comment]
@ericjorgeamaral , @jennifer-shehane  and all, 
Here is an example : 

![image](https://user-images.githubusercontent.com/10066970/72061510-547ba180-32d6-11ea-8a24-3df84be8f478.png)

And my button source code : 
`<button class="root-0-1-21 secondary-0-1-28" id="defaultBtn">Default</button>`

I want to test my color button during the "hover" state and screenshot this state.
If i use it :
` cy.get('#defaultBtn').trigger('mouseenter').matchImageSnapshot();`
or
` cy.get('#defaultBtn').trigger('mouseleave').matchImageSnapshot();`
or
` cy.get('#defaultBtn').trigger('mouseover').matchImageSnapshot();`

For me, it's unable to keep the hover state for checking the background color.

Do you have any workaround ?

-- Env --
Cypress 3.8.1
Chromium 81

**Comment by emirhangl** [comment]
> I am using tippy.js to display tool tips. .trigger('mousehover') does not work.
> Please suggest a way

It's not mousehover, it's mouseover.

**Comment by dudewad** [comment]
@emirhangl `mousehover` is not a DOM mouse event. Try `mouseover` or `mouseenter` and their counterparts, `mouseout` and `mouseleave` respectively.

**Comment by nurkeevich1992** [comment]
Any update on this issue? we started using cypress for our company automated testing. we working with charts and mouse over function very critical for us.

**Comment by AngelRajan** [comment]
Any update on this issue?

**Comment by jennifer-shehane** [comment]
This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.

## Please do not comment asking for updates

Please refrain from commenting asking for updates 🙏 

1. It creates noise for other people watching progress here.
1. It takes the team's focus away from current work.

Our team will comment on this issue when there are any further updates.

**Comment by tacb14** [comment]
Just sharing a workaround that worked for my project, using the right click. I hope this can help some of you.
`cy.get('button').rightclick()`

**Comment by avallete** [comment]
Hi there,

Since Cypress now expose the CDP trough 'remote:protocol' for Chrome/Chromium/Electron browsers.
It allowed me to work on suggestion I did [here](https://github.com/cypress-io/cypress/issues/10#issuecomment-559829533) using CDP to implement a "native hover" in a puppeteer'ish way.

I thought that it may also be of some use to other people so I did make it a package: [cypress-nhover](https://github.com/avallete/cypress-nhover)

It's not perfect (doesn't work with Firefox), but it allowed me to migrate some of my puppeteer tests to cypress pretty easily.

**Comment by bestfei** [comment]
> Just sharing a workaround that worked for my project, using the right click. I hope this can help some of you.
> `cy.get('button').rightclick()`

it works.thx.

**Comment by farhan** [comment]
> This helped me out to at least show my element to be able to click it:
> https://docs.cypress.io/v1.0/docs/invoke#section-properties-that-are-functions-are-invoked
> 
> ```js
> cy.get('.content').invoke('show').click();
> ```

worked like a charm, thanks!

**Comment by rameshrc** [comment]
I have an application that is made of web components with lit-html. I could not make hover to work with the help of invoke or trigger along with experimental shadow dom as true. Any suggestions?

**Comment by dmtrKovalenko** [comment]
It is possible to run a native system hover event using CDP. Find the ready-to-use hover command in this package https://github.com/dmtrKovalenko/cypress-real-events. Check`cy.realHover` command.

**Comment by DannyFeliz** [comment]
> It is possible to run a native system hover event using CDP. Find the ready-to-use hover command in this package [dmtrKovalenko/cypress-real-events](https://github.com/dmtrKovalenko/cypress-real-events). Check`cy.realHover` command.

This worked like a charm 🤗

**Comment by mysticdevx** [comment]
Three workarounds if `.invoke('show')` and `.trigger('mouseover')` doesn't work because of css :hover.
1. `.trigger('mouseenter');`
2. `.rightclick();` 
3. `.nhover();` from here: [https://github.com/avallete/cypress-nhover](https://github.com/avallete/cypress-nhover)

**Comment by vishnuprabhu-95** [comment]
Awesome

**Comment by vegerot** [comment]
> It is possible to run a native system hover event using CDP. Find the ready-to-use hover command in this package https://github.com/dmtrKovalenko/cypress-real-events. Check`cy.realHover` command.

This is really cool.  Unfortunately, I'm looking for a solution that works for Chrome and Firefox :(

**Comment by silversonicaxel** [comment]
@vegerot can't you just target these onHover tests just on Chrome?

**Comment by vegerot** [comment]
@silversonicaxel only if I didn't care about verifying my app on Firefox

**Comment by silversonicaxel** [comment]
Makes totally sense... I've a question by the way @vegerot 
Are you maybe able to take snapshots of this `realHover` event?
I am trying to use Applitools combined with Cypress and this extra plugin, but I can't catch the mouse over state on a snapshot.

Thanks

**Comment by vegerot** [comment]
I haven't tried it

**Comment by dmtrKovalenko** [comment]
@silversonicaxel I think this is impossible

**Comment by prashantabellad** [comment]
https://github.com/dmtrKovalenko/cypress-real-events. works nicely!

**Comment by matthias-dev** [comment]
________________________________
From: prashantabellad ***@***.***>
Sent: Wednesday, May 19, 2021 07:47
To: cypress-io/cypress ***@***.***>
Cc: Matthias Lichtenstein ***@***.***>; Manual ***@***.***>
Subject: [EXTERNAL] Re: [cypress-io/cypress] Add cy.hover() (#10)


https://github.com/dmtrKovalenko/cypress-real-events<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FdmtrKovalenko%2Fcypress-real-events&data=04%7C01%7Cmatthias.lichtenstein%40brz.eu%7Cbc3d0e3823aa4dc10eec08d91a899653%7C7ecb8dff97cb4f4a94d386a4b39a84f2%7C0%7C0%7C637570000516912287%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=udX5vRJdMTjQImmeiLZtyRarUwCIv6NnZmsphy%2F9JN0%3D&reserved=0>. works nicely!

**Comment by DavidGWheeler** [comment]
> https://github.com/dmtrKovalenko/cypress-real-events. works nicely!

Very nice!  I hope this gets implemented for FF as well, but for now, this is excellent!

**Comment by tiec7** [comment]
> > https://github.com/dmtrKovalenko/cypress-real-events. works nicely!
> 
> Very nice! I hope this gets implemented for FF as well, but for now, this is excellent!

I confirm this, I'm using Cypress for job and I have to replicate an hover on a menu.
I spent almost 3 hours, nothing worked but this, using `cy.get(<element>).realHover()` .

Have a try, no big installs, just download, add a line in `cypress/support/index.js` and use it.

**Comment by SalahAdDin** [comment]
> Three workarounds if `.invoke('show')` and `.trigger('mouseover')` doesn't work because of css :hover.
> 
>     1. `.trigger('mouseenter');`
> 
>     2. `.rightclick();`
> 
>     3. `.nhover();` from here: https://github.com/avallete/cypress-nhover thanks to @avallete

I couldn't make it work.

**Comment by dmtrKovalenko** [comment]
You can use cypress-real-events to test css as well.
